### PR TITLE
Update kubevirt builder image to use go1.13.4

### PR DESCRIPTION
**Pull request by danielBelenky**
- Updated the kubevirt builder image to use go1.13.4.
- We now use go modules to get all the dependencies we need.
  It allows us to just go get the specific version that we need instead
  of having to download and compile it ourselves.
- Set the versions of the dependencies statically.
- Updated tests to call `testing.Init()` explicitly in our `init()`
  functions to use the default flags from the testing package. 

Signed-off-by: Daniel Belenky <dbelenky@redhat.com>

Worked on this PR with @xpivarc 

**Release note**:

```release-note
KubeVirt is now being built with Go1.13.4
```

### CODE REVIEW

File: hack/builder/Dockerfile [hunk 1]
```diff
@@ -12,98 +12,61 @@ RUN dnf install -y dnf-plugins-core && \
...
     source /etc/profile.d/gimme.sh && \
-    go get github.com/mattn/goveralls && \
-    go get -u github.com/golang/mock/gomock && \
-    go get -u github.com/rmohr/mock/mockgen && \
-    go get -u github.com/rmohr/go-swagger-utils/swagger-doc && \
-    go get -u github.com/onsi/ginkgo/ginkgo && \
-    cd /go && \
-    go clean -cache -modcache -r && \
-    rm -rf /go && mkdir /go
+    go get -v -u github.com/mattn/goveralls@v0.0.5 && \
```
Comment (xpivarc) [inline hack/builder/Dockerfile:62]: Maybe we can drop -u flag.

File: go.mod [hunk 2]
```diff
@@ -75,7 +76,7 @@ require (
 
 replace (
 	github.com/Azure/go-autorest => github.com/Azure/go-autorest v11.0.0+incompatible
-	github.com/coreos/prometheus-operator => github.com/coreos/prometheus-operator v0.0.0-20181206002233-dd6f23e7207c
+	github.com/coreos/prometheus-operator => github.com/coreos/prometheus-operator v0.28.1-0.20190202022740-dd6f23e7207c
```
Comment (xpivarc) [inline go.mod:79]: Finally my IDE will not complain about this! :+1:

**Comment by danielBelenky** [comment]
> doesn't `hack/builder/version.sh` need to be updated?
> 
> Also in order to merge this, we're going to need the new builder image pushed to the public repo, and for that image's sha to be references in the `hack/dockerized` file for the KUBEVIRT_BUILDER_IMAGE env var.
> 
> This guarantees that our CI is actually testing the new builder image and verifies it works.

Yeah I know, I'm pushing the image now but it takes some time.
Thanks for the heads up :)

File: hack/dockerized [hunk 3]
```diff
@@ -8,7 +8,7 @@ if [ "${KUBEVIRT_RUN_UNNESTED}" == "true" ]; then
     exit $?
 fi
 
-KUBEVIRT_BUILDER_IMAGE="kubevirt/builder@sha256:e6a76ec71f2de87c3bc648bae87ab43a113540bf57da69f07f76bd0edfd500ab"
+KUBEVIRT_BUILDER_IMAGE="kubevirt/builder:30-5.8.1-amd64"
```
[the remark this answers is not in the document]
Comment (danielBelenky) [reply hack/dockerized:11]: Thanks. I've pushed this change by mistake when I ran the verification locally.

File: hack/builder/Dockerfile (hunk 1 above)
Comment (danielBelenky) [reply hack/builder/Dockerfile:62]: ack, thank you

File: go.mod [hunk 4]
```diff
@@ -112,3 +113,5 @@ replace (
 
 	kubevirt.io/client-go => ./staging/src/kubevirt.io/client-go
 )
+
+go 1.13
```
Comment (davidvossel) [inline go.mod:117]: :( sorry, can you remove this. I know your IDE automatically adds it.

File: hack/builder/Dockerfile [hunk 5]
```diff
@@ -112,17 +75,12 @@ RUN \
     go install && \
     cd /test-infra/robots/issue-creator && \
     go install && \
-    cd /test-infra/robots/pr-labeler && \
-    go install && \
-    cd /go && \
-    go clean -cache -modcache -r && \
-    rm -rf /test-infra && \
-    rm -rf /go && mkdir /go
+    cd /test-infra/robots/pr-labeler
```
Comment (davidvossel) [inline hack/builder/Dockerfile:78]: doesn't this need a `go install` commadn as well after entering the directory?

File: hack/builder/Dockerfile [hunk 6]
```diff
@@ -112,17 +75,12 @@ RUN \
     go install && \
     cd /test-infra/robots/issue-creator && \
     go install && \
-    cd /test-infra/robots/pr-labeler && \
-    go install && \
-    cd /go && \
-    go clean -cache -modcache -r && \
-    rm -rf /test-infra && \
-    rm -rf /go && mkdir /go
```
Comment (davidvossel) [inline hack/builder/Dockerfile:120]: why are these lines removed? this should help shrink the image size

File: hack/builder/Dockerfile [hunk 7]
```diff
@@ -12,98 +12,61 @@ RUN dnf install -y dnf-plugins-core && \
...
-    go get -u github.com/rmohr/go-swagger-utils/swagger-doc && \
-    go get -u github.com/onsi/ginkgo/ginkgo && \
-    cd /go && \
-    go clean -cache -modcache -r && \
-    rm -rf /go && mkdir /go
+    go get -v -u github.com/mattn/goveralls@v0.0.5 && \
+    go get -v -u github.com/golang/mock/gomock@v1.4.3 && \
+    go get -v -u github.com/rmohr/mock/mockgen@v0.0.0-20170327080805-5980f5d9a70d && \
+    go get -v -u github.com/rmohr/go-swagger-utils/swagger-doc@v0.0.0-20181110145135-878f7deb1d17 && \
+    go get -v -u github.com/onsi/ginkgo/ginkgo@v1.12.0
```
Comment (xpivarc) [inline hack/builder/Dockerfile:67]: same here

File: hack/builder/Dockerfile [hunk 8]
```diff
@@ -12,98 +12,61 @@ RUN dnf install -y dnf-plugins-core && \
...
-    cd /go && \
-    go clean -cache -modcache -r && \
-    rm -rf /go && mkdir /go
+    go get -v golang.org/x/tools/cmd/goimports@d5fe738 && \
+    go get -v mvdan.cc/sh/v3/cmd/shfmt@v3.1.1 && \
+    go get -v k8s.io/code-generator/cmd/deepcopy-gen@v0.18.2 && \
+    go get -v k8s.io/code-generator/cmd/defaulter-gen@v0.15.12  && \
+    go get -v k8s.io/kube-openapi/cmd/openapi-gen@30be4d1 && \
+    go get -v github.com/golang/protobuf/protoc-gen-go@1643683 && \
+    go get -v k8s.io/code-generator/cmd/client-gen@v0.16.4
```
Comment (xpivarc) [inline hack/builder/Dockerfile:58]: Can we add `go clean -cache -modcache` to lower the size?

File: hack/builder/Dockerfile (hunk 6 above)
Comment (xpivarc) [reply hack/builder/Dockerfile:120]: We should keep that but withou -r flag and `cd /go`

File: hack/builder/version.sh [hunk 9]
```diff
@@ -1,2 +1,2 @@
-VERSION=30-5.8.1
+VERSION=31
```
Comment (xpivarc) [inline hack/builder/version.sh:1]: I think we want `30- *.` From history I can see that the first number corresponds to fedora version...

File: pkg/virt-launcher/monitor_test.go [hunk 10]
```diff
@@ -38,6 +39,7 @@ import (
 var fakeQEMUBinary string
 
 func init() {
+	testing.Init()
```
Comment (davidvossel) [inline pkg/virt-launcher/monitor_test.go:42]: why do we need this now, but didn't need this in the past?

File: hack/bazel-build.sh [hunk 11]
```diff
@@ -28,6 +28,8 @@ mkdir -p ${CMD_OUT_DIR}/dump
 
 # Build all binaries for amd64
 bazel build \
+    --sandbox_debug \
```
Comment (rmohr) [inline hack/bazel-build.sh:31]: This can be added to `.bazelrc` if someone needs it. Do we really want to commit it?

File: go.mod (hunk 4 above)
Comment (danielBelenky) [reply go.mod:117]: This is added back by `make generate`.
Why we don't want this line in `go.mod`? Isn't is explicitly says that we now require go 1.13 to properly get all the dependencies and work?

File: hack/bazel-build.sh (hunk 11 above)
Comment (danielBelenky) [reply hack/bazel-build.sh]: Added this by mistake... removed now.

**Comment by phoracek** [comment]
@EdDev would you please look into the last failing test here? With Go 1.13.4, TLS was bumped from 1.2 to 1.3 and it seems to break one of the migration tests: test_id:2303. I saw you active on migration PR reviews, so that's why I'm asking. Cheers!

**Comment by xpivarc** [comment]
In TLS 1.3, the client finishes the handshake so we don't get an error. So we need to do something like this: 
`conn, err := tls.Dial("tcp", fmt.Sprintf("localhost:4321%d", i), tlsConfig)
if conn != nil {
	fmt.Fprintf(GinkgoWriter, "%+v\n", conn.ConnectionState())
	b := []byte("AHOJ")
	_, err = conn.Read(b)
	if err != nil {
		fmt.Fprintf(GinkgoWriter, "%+v\n", err.Error())
	}
}
Expect(err).To(HaveOccurred())`

**Comment by oshoval** [comment]
/lgtm
just for poc

**Comment by nunnatsa** [comment]
Go 1.13.4 was released on Nov 1, 2019, while the latest version is 1.13.11

Why not taking the latest 1.13 version, that includes bug and security fixes? I'm not talking about 1.14, but within 1.13, each version is more stable than the previous.

**Comment by davidvossel** [comment]
@EdDev @phoracek @xpivarc 

we need this to move forward. Are be blocked on the TLS issue still?

>Why not taking the latest 1.13 version, that includes bug and security fixes? I'm not talking about 1.14, but within 1.13, each version is more stable than the previous.

@nunnatsa good observation. Lets use the most recent 1.13 version we can.

**Comment by xpivarc** [comment]
> @EdDev @phoracek @xpivarc
> 
> we need this to move forward. Are be blocked on the TLS issue still?
> 
> > Why not taking the latest 1.13 version, that includes bug and security fixes? I'm not talking about 1.14, but within 1.13, each version is more stable than the previous.
> 
> @nunnatsa good observation. Lets use the most recent 1.13 version we can.

I am not aware of it.

**Comment by ormergi** [comment]
Hey we need the for https://github.com/kubevirt/kubevirt/pull/3551
is there any progress?

**Pull request closed without merging**